systemtest/infra_kafka.go

```diff
@@ -0,0 +1,230 @@
+// must accept at least these variables:
+// - namespace (string)
+// - name (string)
+// - topics (list(string))
+// The terraform module should output at least these variables:
+// - deployment_type (string)
+// - kafka_brokers (list(string))
+//
+// If `deployment_type` output is "k8s", `kubectl port-forward` will be run
+// targeting the `PortForwardResource` with `PortForwardMapping`.
+func ProvisionKafka(ctx context.Context, cfg KafkaConfig) error {
+	if cfg.Name == "" {
+		if n := os.Getenv("KAFKA_NAME"); n != "" {
+			cfg.Name = n
+		} else {
+			cfg.Name = "kafka"
+		}
+	}
+	if cfg.Namespace == "" {
+		if n := os.Getenv("KAFKA_NAMESPACE"); n != "" {
+			cfg.Namespace = n
+		} else {
+			cfg.Namespace = "kafka"
+		}
+	}
+	logger().Infof("provisioning Kafka infrastructure with config: %+v", cfg)
+	tf, err := NewTerraform(ctx, cfg.TFPath)
+	if err != nil {
+		return err
+	}
+	if err := tf.Init(ctx, tfexec.Upgrade(true)); err != nil {
+		return err
+	}
+	// Create a dummy topic so the infrastructure awaits until all the required
+	// infrastructure is available. It will be destroyed during the first test.
+	if len(cfg.Topics) == 0 {
+		cfg.Topics = append(cfg.Topics, SuffixTopics("dummy")...)
+	}
+	jsonTopics, err := json.Marshal(cfg.Topics)
+	if err != nil {
+		return err
+	}
+	namespaceVar := tfexec.Var(fmt.Sprintf("namespace=%s", cfg.Namespace))
+	nameVar := tfexec.Var(fmt.Sprintf("name=%s", cfg.Name))
+	topicsVar := tfexec.Var(fmt.Sprintf("topics=%s", jsonTopics))
+	// Ensure terraform destroy runs once per TF path.
+	RegisterDestroy(cfg.TFPath, func() {
+		logger().Info("destroying provisioned Kafka infrastructure...")
+		tf.Destroy(ctx, topicsVar, namespaceVar, nameVar)
+	})
+	if err := tf.Apply(ctx, topicsVar, namespaceVar, nameVar); err != nil {
+		return err
+	}
+	tfOutput, err := tf.Output(ctx)
+	if err != nil {
+		return err
+	}
+	if raw, ok := tfOutput[KafkaBrokersKey]; ok {
+		var brokers []string
+		if err := json.Unmarshal(raw.Value, &brokers); err != nil {
+			return err
+		}
+		if len(brokers) > 0 {
+			SetKafkaBrokers(brokers...)
+		}
+	}
+	dt := string(bytes.TrimSpace(tfOutput[KafkaDeploymentTypeKey].Value))
+	// If PortForward is set to true, and the deployment_type is k8s, start
+	// forwarding traffic to/from kafka to localhost.
+	if cfg.PortForward && dt == `"k8s"` {
+		var err error
+		runOnce.Do(func() {
+			// NOTE(marclop) These assume terraform uses the strimzi operator.
+			if cfg.PortForwardResource == "" {
+				cfg.PortForwardResource = fmt.Sprintf("service/%s-kafka-bootstrap", cfg.Name)
+			}
+			if cfg.PortForwardMapping == "" {
+				cfg.PortForwardMapping = "9093:9093"
+			}
+			err = kubectlForwardKafka(cfg.Namespace,
+				cfg.Name,
+				cfg.PortForwardResource,
+				cfg.PortForwardMapping,
+			)
+		})
+		if err != nil {
+			return err
+		}
+	}
+	logger().Info("Kafka infastructure fully provisioned!")
+	return nil
+}
+
+func kubectlForwardKafka(ns, name, service, mapping string) error {
```

Comment:
Instead of calling kubectl commands what do you think about using the [kubernetes client libraries](https://github.com/kubernetes/client-go/)? If positive then maybe we can do a followup.

Reply:
We could. I didn't want to spend too much time on this beforehand in case we didn't think it's a good idea.